components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -195,6 +198,60 @@ class GeckoEngineSession(
         }
     }
 
+    /**
+     * See [EngineSession.requestPdfToDownload]
+     */
+    override fun requestPdfToDownload() {
+        geckoSession.saveAsPdf().then(
+            { inputStream ->
+                if (inputStream == null) {
+                    return@then GeckoResult<Void>()
+                }
```

Comment:
Could we add some comments to clarify when` inputStream == null` happens? Should we communicate back to the user when this happens?

Reply:
That's a good question. @ohall-m and I were thinking about what to do in the case of a geckoexception where we only log it for now. Is there value for the app to know that it failed? Probably - we could throw some toast with a generic message.

I'll look for a pattern that does this already and try to follow that.

Reply:
We could dispatch an action that app could observe, but the toast alternative could be a viable option too

Reply:
The toast would be a result of providing a mechanism for the app to know about the failure. If that's a potential use case, then we should propagate the error up to the caller.

Reply:
I don't think `inputStream == null` would generally occur, after looking through the GeckoView `saveToPdf`, but is a good defensive idea. 

I think it is possible for there to be an edge case where the `inputStream` has 0 bytes, which might need to be handled too. I imagine it would fail gracefully and make an empty file, but is something to consider.

Reply:
I've been thinking about this for a bit. There are some examples of how to approach this problem similar to downloads which used a `DownloadState` to signify success/failure, however I'd like to approach that in a separate follow-up patch to figure out what that state object should look like and the best way to update it.

I don't believe it should block this feature because the core functionality is still there and this state object design would aid the application only on rare failures. Which isn't to say that we shouldn't do a follow up immediately, but it's something that we can do async with the app team so we can get something out to Nightly sooner.

How does that sound?

Reply:
cc: @Amejia481

Reply:
Yeah, sound like a good idea, we could use `DownloadState.status = FAILED` to find a way to signal the error and reuse the current existing UI.

Reply:
Yeah, exactly. I'm stuck between consider if we should use `DownloadState` or make a new `PdfState` if there are more details we want to provide. Either way, that comes off to me as a separate discussion.

Okay - I'll write the tests for this and move it along, thanks all!

Reply:
Yeah, let discuss in separate to unblock this, and then we can go deeper about this topic :)

Reply:
Filed a follow-up: https://github.com/mozilla-mobile/android-components/issues/12642